pkg/controller/schedulingmanager/controller.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulingmanager
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	corev1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	fedclientset "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned"
+	corev1alpha1client "github.com/kubernetes-sigs/federation-v2/pkg/client/clientset/versioned/typed/core/v1alpha1"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/schedulingpreference"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/pkg/schedulingtypes"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+var SchedulingRegistry = map[string][]string{
+	schedulingtypes.RSPKind: {
+		"FederatedDeployment",
+		"FederatedReplicaSet",
+	},
+}
+
+type SchedulerController struct {
+	// Store for the FederatedTypeConfig objects
+	store cache.Store
+	// Informer for the FederatedTypeConfig objects
+	controller cache.Controller
+
+	worker util.ReconcileWorker
+
+	scheduler map[string]schedulingtypes.Scheduler
+
+	config           *restclient.Config
+	fedNamespace     string
+	clusterNamespace string
+	targetNamespace  string
+	minimizeLatency  bool
+	stopChan         <-chan struct{}
+}
+
+func StartSchedulerController(config *restclient.Config, fedNamespace, clusterNamespace, targetNamespace string, stopChan <-chan struct{}, minimizeLatency bool) error {
+
+	userAgent := "SchedulerController"
+	restclient.AddUserAgent(config, userAgent)
+	client := fedclientset.NewForConfigOrDie(config).CoreV1alpha1()
+
+	controller, err := newController(client, config, fedNamespace, clusterNamespace, targetNamespace, stopChan, minimizeLatency)
+	if err != nil {
+		return err
+	}
+	glog.Infof("Starting scheduler controller")
+	controller.Run(stopChan)
+
+	return nil
+}
+
+func newController(client corev1alpha1client.CoreV1alpha1Interface, config *restclient.Config, fedNamespace, clusterNamespace, targetNamespace string, stopChan <-chan struct{}, minimizeLatency bool) (*SchedulerController, error) {
```

Comment:
I found several controllers, e.g.servicedns.StartController, federatedtypeconfig.StartController, have similar code. I guess it is there for further enhancement because main() logic has error checking for related fail condition too.

Reply:
This is an example of YAGNI (You Ain't Gonna Need It).  If no error is currently being returned, the signature should reflect that, and an error can be added back in the future _if_ it becomes necessary.